src/users.rs

```diff
@@ -0,0 +1,254 @@
+use argon2::{self, Config};
+use bytes::Bytes;
+use data_encoding::HEXUPPER;
+use failure;
+use rand;
+use ring::digest;
+use rusqlite::NO_PARAMS;
+use rusqlite::{params, Connection};
+use std::env;
+
+// User Struct
+pub struct User {
+    pub name: String,
+    pub salt: String,
+    pub hash: String,
+    pub active: i32,
+}
+
+// Constants
+const CREDENTIAL_LEN: usize = digest::SHA512_OUTPUT_LEN;
+
+// Error Used for User related Functions
+// Implements the traits of Error
+
+// Generates a salt
+fn get_salt() -> Vec<u8> {
+    let salt: Vec<u8> = (0..CREDENTIAL_LEN).map(|_| rand::random::<u8>()).collect();
+    salt
+}
+
+// Takes in a salt and password and returns the hash
+// using the argon 2 algorithm
+fn get_hash_from_password(password: String, salt: Vec<u8>) -> Result<String, failure::Error> {
+    let config = Config::default();
+    let hash = argon2::hash_encoded(password.as_bytes(), &Bytes::from(salt), &config)?;
+    Ok(hash)
+}
+
+// Verifies a password and salt against a hash
+// using the argon 2 algorithm
+fn verify_password(password: String, hash: String) -> Result<bool, failure::Error> {
+    let valid = argon2::verify_encoded(&hash, password.as_bytes())?;
+    Ok(valid)
+}
+
+// creates a sqlite DB if it does not exist and initializes the user table
+// returns the connection
+fn connection() -> Result<rusqlite::Connection, failure::Error> {
+    let db_file = env::var("DB_FILE").unwrap_or("sqlite.db".to_string());
+    let conn = Connection::open(db_file).expect("db conn fail");
+
+    // create the users table if it does not exist
+    conn.execute(
+        "create table if not exists users (
+             id integer primary key,
+             name text not null unique,
+             salt text not null,
+             hash text not null,
+             active integer default 1 not null
+         );",
+        NO_PARAMS,
+    )?;
+    Ok(conn)
+}
+
+impl User {
+    // trait used to create a user
+    pub fn create(&mut self, password: String) -> Result<(), failure::Error> {
```

Comment:
I'm thinking this should maybe change to `new(name: String, password: String) -> Result<User, Error> { ...`

Then take the `pub` off the struct members. I don't _think_ the caller is interested in setting anything except name and password, right?

Reply:
at this stage I dont beleive so, I will update this ASAP

Reply:
Done, @amouat I updated all the traits to follow this as I felt it was a better pattern